acsylla/_cython/statement/statement.pyx

```diff
@@ -93,138 +95,134 @@ cdef class Statement:
-        cdef CassError error
-        error = cass_statement_bind_float_by_name_n(
-            self.cass_statement, name, len(name), value)
-        raise_if_error(error)
+    cdef const CassDataType* _get_data_type_by_name(self, bytes name):
+        cdef const CassDataType* data_type
 
-    cdef inline _bind_bool_by_name(self, bytes name, object value):
-        cdef CassError error
-        if value is True:
-            error = cass_statement_bind_bool_by_name_n(
-                self.cass_statement, name, len(name), cass_true)
+        if self.cass_prepared:
+            data_type = cass_prepared_parameter_data_type_by_name(self.cass_prepared, name)
+            return data_type
         else:
-            error = cass_statement_bind_bool_by_name_n(
-                self.cass_statement, name, len(name), cass_false)
-        raise_if_error(error)
-
-    cdef inline _bind_string_by_name(self, bytes name, str value):
-        cdef CassError error
-        cdef bytes bytes_value
-
-        bytes_value = value.encode()
-
-        error = cass_statement_bind_string_by_name_n(
-            self.cass_statement, name, len(name), bytes_value, len(bytes_value))
-        raise_if_error(error)
-
-    cdef inline _bind_bytes_by_name(self, bytes name, bytes value):
-        cdef CassError error
-        error = cass_statement_bind_bytes_by_name_n(
-            self.cass_statement,
-            name,
-            len(name),
-            <const cass_byte_t*> value,
-            len(value)
-        )
-        raise_if_error(error)
-
-    cdef inline _bind_uuid_by_name(self, bytes name, TypeUUID uuid):
-        cdef CassError error
-        cdef CassUuid cass_uuid
-        cdef bytes bytes_value
-
-        bytes_value = uuid.uuid.encode()
+            raise ValueError(
+                "Method only availabe for statements created from prepared statements")
 
-        cass_uuid_from_string(bytes_value, &cass_uuid)
-        error = cass_statement_bind_uuid_by_name_n(
-            self.cass_statement,
-            name,
-            len(name),
-            cass_uuid
-        )
-        raise_if_error(error)
+    cdef CassValueType _get_value_type(self, const CassDataType* data_type):
+        cdef CassValueType cass_value_type
+        cass_value_type = cass_data_type_type(data_type)
+        return cass_value_type
 
     cpdef bind(self, int idx, object value):
+        from uuid import UUID
+        from decimal import Decimal
+        from ipaddress import IPv4Address, IPv6Address
+        from datetime import date, datetime, time, timedelta
+        cdef const CassDataType* cass_data_type
+        cdef CassValueType cass_value_type
+
+        try:
+            bind_null(self.cass_statement, idx)
+        except CassErrorLibIndexOutOfBounds:
+            raise CassErrorLibIndexOutOfBounds()
 
         if value is None:
-            self._bind_null(idx)
+            return
+
+        if self.cass_prepared:
```

Comment:
do we need to test if its prepared or not? I would say that we can use the `bind_xxx(self.cass_statement, ....)` without considering if its preared or not.

Or am I missing something? do we need to get the data type using the `_get_data_type`? does this allow us to fail early?

Reply:
Yes we need this check because only for prepared statement we can get value type for column. For non prepared statement we map value by python type. 

I think there is an alternative way if we use the column metadata for example cass_column_meta_data_type then this check will not be needed